Slot mappings documentation update (#10011)
RasaHQ/rasa: docs/docs/forms.mdx

### Slot Mappings

:::caution Changed in 3.0

> As of 3.0, [slot mappings](./domain.mdx#slot-mappings) can only be defined in the slots section of the domain.
> Please follow the migration guide to update your assistant.

Reviewer: ```suggestion
As of 3.0, [slot mappings](./domain.mdx#slot-mappings) are defined in the `slots` section of the domain.
Please follow the migration guide to update your assistant.
```

Since there wasn't any alternative before it's not a limitation, just a move.

Reviewer 2: @ancalita how about a short sentence that says that now it's deterministic, as opposed to automatic/guessing in 2.x + link to migration guide?

:::

#### Unique entity mapping matching

In `from_entity` mapping, when an extracted entity uniquely maps onto a slot,

> the slot will be filled even if this slot wasn't requested by the form.

Reviewer: I know you didn't write this, but could you add a short explanation/example of what "uniquely maps" means here?

### Slot Mappings

:::caution Changed in 3.0
As of 3.0, [slot mappings](./domain.mdx#slot-mappings) are defined in the `slots` section of the domain.
This change allows the same slot mapping to be reused across multiple forms, removing any unnecessary duplication.
Please follow the [migration guide](./migration-guide.mdx#slot-mappings) to update your assistant.
:::

#### Unique entity mapping matching

> In `from_entity` mapping, when the attributes of an extracted entity (i.e, entity name, role, group) uniquely map

Reviewer: ```suggestion
In the `from_entity` slot mapping, when the attributes of an extracted entity (i.e, entity name, role, group) uniquely map
```

Reviewer: Also: What is "uniquely" - the slot has only this mapping from this entity, or the entity is only referenced by one slot mapping?

Author: The slot can have multiple mappings, and some can be shared by other slots. But if the slot contains a mapping that is not found in other slots' mappings, then we consider it unique.  Or the entity can be referenced in multiple mappings, but if the role/group is only specific to a slot, then the mapping is unique to that slot.

> onto a slot, the slot will be filled even if this slot wasn't requested by the form.

Reviewer: isn't this the case for all slot mappings? They will apply even if the slot was not requested excepting conditions 🤔
This concept was kinda confusing in 2.x too, but now it's more so because slot filling happens somewhere else.

Is this only applicable when `store_entities_as_slots` is on? I think I need to understand this behaviour before suggesting anything 😅

Author: This is actually separate from auto-fill (which required `store_entities_as_slots`). 

To maintain regression of this unique entity mapping feature in forms, the new slot mappings maintains it during an active_loop with slot mappings that have conditions. So it will not extract the slot if the user utterance contained an entity which is mapped onto a non-unique `from_entity` mapping and the slot was not requested by the form.

Reviewer: Let me see if I've got this right:
1. The definition of "unique entity mapping":

There is only one slot mapping in the entire domain which is `from_entity`  for a given (entity + attributes)  combo . 

2. Applicability of `from_entity` mappings when a form is active

`from_entity` mappings will get applied while a form is active iff either 
(1) the form requested the slot OR
(2) There is only one slot in the entire domain with a slot mapping for the specific (entity +attributes) combo listed.
 
Is this correct?

Reviewer: I remember the point of this now (also based on example below), which was to prevent filling a slot with an entity intended to fill only the requested slot e.g. given a form that collects both `amount` (from `number` entity) and `account_number` (also from `number` entity), the form shouldn't go filling both slots with the first `number` entity extracted, it should leave the not-yet-requested one empty and wait for the user's response to the specific prompt.

If my understanding above is correct, I'll draft a short explanation of what "unique" means, because I'm sure someone besides me is gonna wonder about it

Author: > Applicability of from_entity mappings when a form is active

There is a nuance here when the form is active:
- either a slot is requested -> any of its mappings (unique or non-unique) become eligible for filling that slot.
- a slot was not requested but the user utterance has an entity that could fill that slot -> that entity can fill that slot only according to a unique entity mapping; otherwise a non-unique entity mapping won't apply and the slot remains unfilled.

Reviewer: Right, I think we're on the same page now 😅 Let me make a phrasing suggestion then.

Author: Hmm, it's intended to allow other slots that are not in that form's required_slots list to be filled, if those slots don't have mapping with conditions (we assumed that any slot without conditions is meant to be filled by any user utterance outside forms too).

> it will still fill slots required by other forms with an extracted entity even if a form is active and the mapping is not unique across all slots in the domain.

Let's say you have slot_2 that should only be set by form_2, slot_2 should have its mapping with conditions to prevent this behaviour you mention in form_1 that doesn't require slot_2.

Maybe something else I need to clarify in the docs in `mapping conditions` section?

Reviewer: I think this is the right place to explain it, will include in update I'm pushing

1. Define a method `extract_<slot_name>` for every slot that should be mapped in a
   custom way.

> 2. Make sure that in the domain file you list for your form only those slots that use 

Reviewer: ```suggestion
2. In your domain file, for your form's `requested_slots`, list only those slots that use
```

Is this still true? Can it not list slots with custom mappings under `required_slots` now?

Author: Both ways should work actually but decided to keep the old docs in since it's not necessary to migrate 2.x `FormValidationAction` to the new way - I've made a note to test this today actually to make sure both cases work and let you know.

Author: Yes, you can list slots with custom mappings under form `required_slots`, meaning you don't have to override the `required_slots` anymore, and it actually gives the extra benefit that you can use `utter_ask_<custom_slot>` for these slots now. However, you can continue implementing the 2.x way currently described in the docs too. How should we write that both ways work?

Reviewer: Great that both ways do work! 
Since this is not the migration guide, my instinct is it should rather push users to the most sensible usage in 3.0, which is listing all required slots, custom and otherwise, in the domain, and only indicate the possibility of overwriting the method in the custom action when actual dynamic behaviour is required i.e. in the `dynamic form behaviour` below.
